FTP uploader plugin

moix opened the pull request:
This simple plugin add the capability to upload output to a ftp

soulgalore commented:
Hey @moix thanks for the PR! 

A standard ftp-plugin is a really good idea!

It's the linting that fails: https://travis-ci.org/sitespeedio/sitespeed.io/builds/297368274?utm_source=github_status&utm_medium=notification

Can you fix that and I'll have a look? `npm run lint` and best case `npm run lint:fix`.

Best
Peter

moix commented:
hey @soulgalore thanks! sorry I'm really a newbie here, can you point me how to setup npm lint? eslint fails due missing conf file. I tried `eslint --init` but dunno what options should put
Best /moix

soulgalore commented:
Hey @moix hmm I wonder if something wrong in our setup, ca share the exact error you get (ping @tobli who has better knowledge of how it should be done).

@moix I think this is a really good PR, I'll just add a couple of things that need to be fixed to follow best practices ok?

soulgalore commented on lib/plugins/ftp/index.js:
```diff
@@ -0,0 +1,75 @@
+  user: 'user',
+  password: 'password'
+};
+function createFTPClient(ftpOptions) {
```
Is this function used anywhere, i cannot find it? The ftpOptions is already defined in the function body as const, you can just rename it.

soulgalore commented on lib/plugins/ftp/index.js:
```diff
@@ -0,0 +1,75 @@
+      const ftpOptions = merge({}, defaultConfig, this.ftpOptions);
+      const baseDir = this.storageManager.getBaseDir();
+      const storagePathPrefix = this.storageManager.getStoragePrefix();
+      var client = new PromiseFtp();
```
All var can be const, so just change all vars to be const :)

soulgalore commented on lib/plugins/ftp/index.js:
```diff
@@ -0,0 +1,75 @@
+        var filelist = [];
+
+        function walkSync(dir) {
+          var path = path || require('path');
```
We usually do all requires at the top of the file to make it easy to see what's used, so if you can move it, thanks!

soulgalore commented on lib/plugins/ftp/index.js:
```diff
@@ -0,0 +1,75 @@
+
+        function walkSync(dir) {
+          var path = path || require('path');
+          var fs = fs || require('fs'),
```
We have fs-extra already we can use that one right?

soulgalore commented on lib/plugins/ftp/index.js:
```diff
@@ -0,0 +1,75 @@
+  password: 'password'
+};
+function createFTPClient(ftpOptions) {
+  const ftpOptions = merge({}, defaultConfig, options.ftp);
```
ftpOptions is already defined as the parameter.

moix commented:
Hi @soulgalore thanks for your comments about good practices, I have added them.
@tobli any other suggestion about what can be failing during build? there was a push with some lint failures [build 297866565](https://travis-ci.org/sitespeedio/sitespeed.io/builds/297866565?utm_source=github_status&utm_medium=notification), I fixed them but now is failing with initial failure we had yesterday [build 297872371](https://travis-ci.org/sitespeedio/sitespeed.io/builds/297872371?utm_source=github_status&utm_medium=notification)

tobli commented:
Prettier is run as a eslint plugin. You first need to run npm install to install the versions of eslint an prettier that's defined in package.json. Try this:
```
# install dependencies (eslint, prettier etc.)
npm install
# check (and fix) code style
npm run lint:fix
```

You'll see one eslint error for an unused 'resolve' function, i.e. there's a promise that's never resolved. Hence the message processing will hang in the queue.
